### PR TITLE
Turn the new lock file format on by default

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit enables the support added in #7070 by default. This means
that gradually over time all `Cargo.lock` files will be migrated to the
new format. Cargo shipped with Rust 1.38.0 supports this new lock file
format, so any project using Rust 1.38.0 or above will converge quickly
onto the new lock file format and continue working.

The main benefit of the new format is to be more friendly to git merge
conflicts. Information is deduplicated throughout the lock file to avoid
verbose `depedencies` lists and the `checksum` data is all listed inline
with `[[package]]`. This has been deployed with rust-lang/rust for some
time now and it subjectively at least seems to have greatly reduced the
amount of bouncing that happens for touching `Cargo.lock`.

### CODE REVIEW

**Comment by ehuss** [comment]
So I imagine this will be problematic for projects that have extra long backwards compatibility.  Is it correct that this is a forced upgrade for everyone?

What about having it so that it only applies to new projects for a while?  That is, it only uses the new format if `Cargo.lock` is missing.  That might allow more people to start using it now without disrupting projects with longer compatibility concerns.

**Comment by alexcrichton** [comment]
Yeah that's true that this would be an upgrade for everyone, but it notably only affects projects which span many toolchain versions and are also actively developed on all those versions. While that seems pretty rare to me, I think you've got a good idea to start out by just phasing in with new lock files, so I'll see if I can whip that up.

**Comment by alexcrichton** [comment]
Ok updated!

**Comment by ehuss** [comment]
@alexcrichton the macos test failure looks legitimate.  I haven't been able to fully dig into it, but it looks like it is rebuilding when it shouldn't.  I can maybe try to dig more tonight.

**Comment by ehuss** [comment]
@alexcrichton Here's a repro (any platform):

1. `cargo new foo`
2. Add a `build.rs` with `fn main() {}`
3. `cargo build`
4. `cargo build` again.  The second one will rebuild when it should be fresh.

The issue is that the `Cargo.lock` is getting rewritten with an extra blank line.  I believe this is because `cargo new` creates a v2 lock, then `cargo build` reverts it to v1 (because there are no dependencies) which adds the newline.  The build script is necessary to make Cargo notice the timestamp change on Cargo.lock.

I'm a bit surprised this didn't trip up any other tests.

**Comment by alexcrichton** [comment]
Ok I've added a new test and pushed up what I believe should be a fix for this, but it's a pretty targeted fix and I don't have a ton of confidence in it. Next time we change formats I think we should do so via https://github.com/rust-lang/cargo/issues/7144 since that'll make this much easier to deal with.

**Comment by ehuss** [comment]
Seems good enough to me.

@bors r+

File: src/cargo/core/resolver/resolve.rs [hunk 1]
```diff
@@ -59,9 +59,12 @@ pub struct Resolve {
 ///
 /// It's theorized that we can add more here over time to track larger changes
 /// to the `Cargo.lock` format, but we've yet to see how that strategy pans out.
-#[derive(PartialEq, Clone, Debug)]
+#[derive(PartialEq, Eq, Clone, Copy, Debug, PartialOrd, Ord)]
 pub enum ResolveVersion {
```
Comment (mathstuf) [inline src/cargo/core/resolver/resolve.rs]: The doccomment for this still says `V1` is the default.

**Comment by ehuss** [comment]
> Is there some way to continue using the old format?

If you have the old format, it will continue to be used.  If you deleted the lock file, you can use an older version of Cargo to recreate it.

File: src/cargo/core/resolver/resolve.rs (hunk 1 above)
Comment (ehuss) [reply src/cargo/core/resolver/resolve.rs]: #7637

**Comment by mathstuf** [comment]
No, I can't, because I generate the file with nightly to get `-Z minimal-versions` support. But, I want to test with my MSRV. Now the `Cargo.lock` is unusable for this setup.

**Comment by mathstuf** [comment]
I guess to clarify, this is a library, so there's no `Cargo.lock` committed.

**Comment by mathstuf** [comment]
And if a `Cargo.lock` *is* committed, I found it hard to justify testing `-Z minimal-versions` anyways; anyone using it isn't getting the version creep that `-Z minimal-versions` will help to avoid in the first place.

**Comment by ehuss** [comment]
Unfortunately we are [recommending](https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#minimal-versions) that people discontinue using `minimal-versions`, as we don't believe the design is going to work out.

**Comment by mathstuf** [comment]
Ok…why was that not announced on the tracking issue? (https://github.com/rust-lang/cargo/issues/5657) In any case, having it available today allows folks to go and find the projects that won't support the shallow minimal-versions so that the ecosystem *is* better when it lands in stable. And if that effort is wide and effective enough, maybe the deep version will be viable at some point. I do have reservations that things like https://github.com/unicode-rs/unicode-normalization/commit/cd63ded085a1b80bbbd84e4b667214cf75ff7c99 will crop up with a shallow variant forcing a creep of MSRV across the ecosystem.

File: src/cargo/core/resolver/resolve.rs [hunk 2]
```diff
@@ -370,23 +407,26 @@ impl fmt::Debug for Resolve {
...
-    /// This means that this encoding version is not currently the default but
-    /// intended to become the default "soon".
-    pub fn from_the_future(&self) -> bool {
-        match self {
-            ResolveVersion::V2 => true,
-            ResolveVersion::V1 => false,
-        }
+    /// It's important that this trails behind `default_for_new_lockfiles` for
+    /// quite some time. This gives projects a quite large window to update in
+    /// where we don't force updates, so if projects span many versions of Cargo
```
Comment (est31) [inline src/cargo/core/resolver/resolve.rs]: Could there be a config param for this? Somewhere in `Cargo.toml` maybe, or alternatively `.cargo/config`?

Comment (mathstuf) [reply src/cargo/core/resolver/resolve.rs]: A flag to `generate-lockfile` would work nicely too.

**Comment by gwillen** [comment]
My group just hit this, and it seems like the error message that older cargo produces when reading a newer lockfile is particularly unhelpful -- I don't know that anything can be done about this now, but it doesn't seem great in terms of the debugging-time cost for people that encounter it. (Backwards-incompatible changes are easier to deal with if they at least produce obvious failures pointing to the need to upgrade.)

I can't imagine this is really fixable at this point, since the behavior of old cargo can't be changed, and the format can't be adjusted without making a _second_ breaking change, but is there some way this could be improved for the future? I see that _newer_ cargo understands the file to be implicitly versioned, and determines this by its contents. But older cargo of course doesn't understand this and just crashes. For future updates, would an explicit version field help avoid this, so that older cargo can declare that it does not understand the newer format?

(An approach I've seen before, which I think would work well here, would be a "minimum cargo version to parse this file" at the top of the file -- not the version that generated the file, but the first version that supports the format being used. Then older versions could not only error out helpfully, they could even point to what version is required to upgrade to.)

**Comment by ehuss** [comment]
@gwillen Discussion of a version in `Cargo.lock` is in #7144.
I think [RFC 2495](https://github.com/rust-lang/rfcs/blob/master/text/2495-min-rust-version.md) would implement something along the lines of a minimum version declaration for `Cargo.toml` which would be an implicit minimum for `Cargo.lock`, though that would be opt-in, and is not yet implemented.

**Comment by kentfredric** [comment]
@mathstuf I have duct taped together some utterly terrible code, that will mangle a new style Cargo.lock contents back into an old style lock file in "Cargo.lock".

Its presented as-is in the hope its useful to somebody.

It may be a bit messy to read given how I've had to implement parts of a statemachine and various parser grammar in-line, so ... good luck extending it.

Regex ninja skills may be required, but all the regex is *relatively* simple, but in ways that mean anything weird I didn't see in my own lock files may not be dealt with at all.

The logic is doable in rust, but I wanted to spend less than an hour making something work.

Enjoy? Feel free to hate the perl.

( I had initially embarked on hand-transcribing a lockfile in a text editor before I decided it would be faster and more reliable to script it poorly )

```perl
#!perl
use strict;
use warnings;

# This is very crude.
# 
# 1. Read ./Cargo.lock trying to extract crate name/version/checksum/source details from 'new' format
# 2. Read Cargo.lock a second time, writing it to Cargo.lock.new streamed, but
#    a. strip checksum = from [[ package ]] entries on the way through
#    b. rewrite dependency = [ sections using data collected on the first pass
# 3. Then, generate a [metadata] section from the data obtained in the first pass and emit to the new file.
#
# Its up to you to copy the new file back into place.

my $states = parse_meta("Cargo.lock");
rewrite_file("Cargo.lock","Cargo.lock.new", $states);
1;

sub fix_dep {
    my ( $depstring, $states ) = @_;
    if ( $depstring =~ /^([^ ]+)$/ ) {

        # single map
        my $package = $1;
        if ( not exists $states->{$package} ) {
            die "$package not found";
        }
        my ( $first, @garbage ) = values %{ $states->{$package} };
        if (@garbage) {
            die "Too many versions for $package";
        }
        return
            $package . " "
          . $first->{version} . " ("
          . $first->{source} . ")";
    }
    if ( $depstring =~ /^([^ ]+)[ ]+([^ ]+)$/ ) {
        my ( $package, $version ) = ( $1, $2 );
        if ( not exists $states->{$package} ) {
            die "$package not found";
        }
        if ( not exists $states->{$package}->{$version} ) {
            die "$package v$version not found";
        }
        return
            $package . " "
          . $version . " ("
          . $states->{$package}->{$version}->{source} . ")";
    }
    die "Unhandled depstring $depstring";
}

sub parse_meta {
    my ($filename) = @_;
    open my $fh, "<", $filename or die "Can't read $filename";

    my $states = {};

  lines: while ( my $line = <$fh> ) {
        next unless $line =~ /^\[\[package\]\]/;
        my $record;

      fields: while ( my $line = <$fh> ) {
            last fields if $line =~ /^\s*$/;
            if ( $line =~ /^([^ ]+) = "([^"]+)"/ ) {
                my ( $field, $value ) = ( $1, $2 );
                $record->{$field} = $value;
            }
        }
        if ( not exists $record->{name} ) {
            die "Didn't parse record name";
        }
        if ( not exists $record->{version} ) {
            die "Didn't parse record version for $record->{name}";
        }
        if ( not exists $record->{source} ) {
            warn
"Didn't parse record source for $record->{name} $record->{version}";
        }
        if ( not exists $record->{checksum} ) {
            warn
"Didn't parse record checksum for $record->{name} $record->{version}";
        }
        $states->{ $record->{name} }->{ $record->{version} } = $record;
    }
    return $states;
}

sub rewrite_file {
    my ( $src, $dest, $states ) = @_;
    open my $fh,  "<", $src  or die "Can't read $src";
    open my $wfh, ">", $dest or die "Cant write $dest";

  lines: while ( my $line = <$fh> ) {
        if ( $line !~ /^\[\[package\]\]/ ) {
            $wfh->print($line);
            next lines;
        }
        $wfh->print($line);
      fields: while ( my $line = <$fh> ) {
            if ( $line =~ /^\s*$/ ) {
                $wfh->print($line);
                last fields;
            }

            if ( $line =~ /^([^ ]+) = "/ ) {
                if ( "$1" ne "checksum" ) {
                    $wfh->print($line);
                    next fields;
                }
            }
            if ( $line =~ /^dependencies = \[/ ) {
                $wfh->print($line);
                dep: while ( my $dep = <$fh> ) {
                    if ( $dep =~ /^\]/ ) {
                        $wfh->print($dep);
                        next fields;
                    }
                    if ( $dep =~ /^(\s+")([^"]+)(".+\z)/ms ) {
                        my ( $lpad, $content, $rpad ) = ( $1, $2, $3 );
                        $wfh->print(
                            $lpad . fix_dep( $content, $states ) . $rpad );
                        next dep;
                    }
                    die "Unhandled $dep";
                }
            }
        }
    }
    # write a metadata section
    $wfh->print("\n[metadata]\n");
    for my $crate ( sort keys %{$states} ) {
      for my $version ( sort keys %{$states->{$crate}} ) {
        next unless exists $states->{$crate}->{$version}->{checksum};
        $wfh->printf(qq["checksum %s %s (%s)" = "%s"\n],
          $crate,
          $version,
          $states->{$crate}->{$version}->{source},
          $states->{$crate}->{$version}->{checksum});
      }
    }
}

1;
```

**Comment by est31** [comment]
Btw, there is also the cargo-lock crate which can translate between formats: https://github.com/rustsec/cargo-lock

**Pull request merged**